substrate/runtime/staking/src/lib.rs

```diff
@@ -381,6 +494,114 @@ impl<T: Trait> Module<T> {
+		let try_index: usize = index.as_();
+		let try_set = Self::enum_set(T::AccountIndex::sa(try_index / ENUM_SET_SIZE));
+		let item_index = try_index % ENUM_SET_SIZE;
+		item_index < try_set.len() && Self::voting_balance(&try_set[item_index]).is_zero()
+	}
+
+	/// Register a new account (with existential balance).
+	fn new_account(who: &T::AccountId, balance: T::Balance) -> NewAccountOutcome {
+		let next_set_index = Self::next_enum_set();
+
+		// A little easter-egg for reclaiming dead indexes..
+		let ret = {
+			let next_set_index: usize = next_set_index.as_();
+
+			// we quantise the number of accounts so it stays constant over a reasonable
+			// period of time.
+			const QUANTIZATION: usize = 256;
+			let quantized_account_count = (next_set_index * ENUM_SET_SIZE / QUANTIZATION + 1) * QUANTIZATION;
+			// then modify the starting balance to be modulo this to allow it to potentially
+			// identify an account index for reuse.
+			let maybe_try_index = balance % T::Balance::sa(quantized_account_count * 256);
+			let maybe_try_index = As::<T::AccountIndex>::as_(maybe_try_index);
+			let maybe_try_index = As::<usize>::as_(maybe_try_index);
+
+			// this identifier must end with magic byte 0x69 to trigger this check (a minor
+			// optimisation to ensure we don't check most unintended account creations).
+			if maybe_try_index % 256 == RECLAIM_INDEX_MAGIC {
+				// reuse is probably intended. first, remove magic byte.
+				let try_index = maybe_try_index / 256;
+
+				// then check to see if this balance identifies a dead account index.
+				let set_index = T::AccountIndex::sa(try_index / ENUM_SET_SIZE);
+				let mut try_set = Self::enum_set(set_index);
+				let item_index = try_index % ENUM_SET_SIZE;
+				if item_index < try_set.len() {
+					if Self::voting_balance(&try_set[item_index]).is_zero() {
+						// yup - this index refers to a dead account. can be reused.
+						try_set[item_index] = who.clone();
+						<EnumSet<T>>::insert(set_index, try_set);
+
+						return NewAccountOutcome::GoodHint;
+					}
+				}
+				NewAccountOutcome::BadHint
+			} else {
+				NewAccountOutcome::NoHint
+			}
+		};
+
+		// insert normally as a back up
+		let mut set_index = next_set_index;
+		// defensive only: this loop should never iterate since we keep NextEnumSet up to date later.
+		let mut set = loop {
+			let set = Self::enum_set(set_index);
+			if set.len() < ENUM_SET_SIZE {
+				break set;
+			}
+			set_index += One::one();
+		};
+
+		// update set.
+		set.push(who.clone());
+
+		// keep NextEnumSet up to date
+		if set.len() == ENUM_SET_SIZE {
+			<NextEnumSet<T>>::put(set_index + One::one());
+		}
+
+		// write set.
+		<EnumSet<T>>::insert(set_index, set);
+
+		ret
+	}
+
+	/// Kill an account's free portion.
+	fn on_free_too_low(who: &T::AccountId) {
+		<FreeBalance<T>>::remove(who);
+		<Bondage<T>>::remove(who);
+		<CodeOf<T>>::remove(who);
+		// TODO: <StorageOf<T>>::remove_prefix(address.clone());
```

Comment:
is there any storage in particular that this would target? IMO it seems dangerous to have the staking module arbitrarily clear storage from any other modules, whatever they may be

Reply:
It's used only for smart contracts which the staking module has sole ownership and operation rights over.